### ISSUE TITLE
Is it OK to include images with no labels in the training data set?

### ISSUE THREAD

**Issue by tischi** [ultralytics/ultralytics#7981]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi,

We have a dataset where some images do not contain any objects at all; and it is in fact important for us that they do not contain objects; so it is important for us that YOLO would not find any objects on these images. 

My question is whether to include images without objects into the training? 

Does it (i) harm the training to include images without objects, or (ii) would YOLO just ignore those images during the training, or (iii) (ideal for us) would it even learn from these images that it should not find objects there? 

Thank you very much for your help!

**Comment by Burhan-Q** [comment]
Images without objects (I refer to them as null images) are not harmful to include, it would _only_ be harmful if the images contained objects without annotations. Null images help reduce the likelihood of false positive detections but otherwise won't provide much help with the model performance. General rule of thumb is to include up to 10-20% null images if there are issues with false positive detections, but remember adding images will increase time for training.

Annotation files are not required to be included with null images, however I tend to create empty text files for null images when I do training. My rationale is that it makes it simple to check 1-to-1 that there are the same number of annotation files as images and that every image has a matching/corresponding annotation file (limits the need for searching).

**Comment by tischi** [comment]
@Burhan-Q thank you very much for your insights!

This brings me to a related question: I have some project where there are very many objects of the same type in one image. Do I need to annotate **all of them** for YOLO to find all of them or would it learn from a few positive examples? In other words, if I do _not_ annotate certain objects, does that make YOLO learn that those are not objects? From your comment above I think I would conclude this, because otherwise NULL images should not help with false positives, correct?

**Comment by Burhan-Q** [comment]
@tischi I generally answer this question with an analogy. If you are learning something, but only are told 5% of the time when you're correct or incorrect, it will be nearly impossible to learn that task/subject. Neural network models are similar in that they must be given the input for corrective actions, otherwise they will never "learn" what an object is.

Humans have the ability to extrapolate with fuzzy logic quickly, but machines do not. Machines are only able to execute complex pattern recognition, but can only do so with a sufficient amount of correct data.

In summary, it's good practice to label/annotate _every_ instance of objects you aim to use your model to detect with. Failing to do so will result in a poorly trained model as unlabeled data is considered "background" or "null" samples. There are caveats and exceptions to this that I would consider edge-cases, and I would advise that **all objects are labeled. Always.**

**Comment by tischi** [comment]
Thank you @Burhan-Q a lot for your answer, which makes sense to me!

**Comment by abhiagwl4262** [comment]
As much as I know, ultralytics filters images with no annotations. Is that correct ? If yes, then why ?

**Comment by glenn-jocher** [comment]
Hello!

Yes, by default, Ultralytics YOLO models skip images without annotations during training. This is primarily to ensure that every image used in training contributes directly to learning, optimizing both the efficiency and effectiveness of the training process. If an image without objects (null image) is meant to contribute to training (like teaching the model where objects aren't), it should still have an associated annotation file, just without any object listings.

If you want to include images with no annotations and ensure they are not filtered out, you might consider adding an empty annotation file for each such image. This way, the model understands these are intentional parts of the dataset.

Hope this clears things up! 🌟

**Comment by abhiagwl4262** [comment]
@glenn-jocher I will see if this works. Thanks for the clarification.

**Comment by glenn-jocher** [comment]
@abhiagwl4262 you're welcome! If you have any more questions as you work through it, feel free to ask. Happy coding! 😊

**Comment by MuhammadShifa** [comment]
Hi ### @glenn-jocher and ### @Burhan-Q ,
Hope to find you well.
From the above discusson we can get that if we have an image without annotation, we still need to add empty text file to have better accuracy and efficiency of yolov8 model.
Now my question is, Can we kept the the label as a single "object" class for multiple classes. For example if I want to detect the coco labels detection generally as an "object", as we have also "SKU-110K" dataset?
what would be the efficiency and accuracy of Yolov8 model in such situtation where a different objects are assigned as a single label "object". Please provide some insights. Thanks

**Comment by glenn-jocher** [comment]
@MuhammadShifa hi there!

Absolutely, you can label multiple object classes as a single "object" class if your application doesn't require distinguishing between them. This approach simplifies the model and can be particularly useful in scenarios where you're only interested in detecting the presence of objects, regardless of their type.

However, treating all objects as a single class may affect the model's ability to learn specific features of different objects, which could potentially reduce its overall accuracy and precision for tasks where such distinctions are important.

For a general presence detection task, this method should work efficiently, but the performance really depends on the specifics of your use case and the diversity of the objects in your dataset.

Hope this helps! 😊

**Comment by pvtoan** [comment]
Hi @glenn-jocher ,

Thank you for your clear explanations about how to use null images in this current version.

However, I have some confuse based on your previous statement for Yolo-v5 in https://github.com/ultralytics/yolov5/issues/2844
In that v5, there is no need to create empty label files for null images. Also, the number of null images should be up to 10%

Therefore, 
1. I would like to know more details what ultralytics version (or which Yolo version) I need to create empty label files and what is not?
Besides, what is the optimal ratio for the number of null images? 0%-10% or 10%-20%?

2. Or, is there any way I can check inside the code to see which ultralytics skips null images without according empty label files during training?
By this way, I can clearly control my training processs as what I need.

3. Applying null images during training can reduce false positive detections for object detection. 
However, is it helpful to applying to the segmentation task?
The reason is that, in my bin picking application, if all objects are overlapped each other (not allowed to pick up) in an image, I cannot label any segmentation there. That means, I also want to teach my model in which cases it cannot segment any object.

**Comment by glenn-jocher** [comment]
Hello,

1. For YOLOv5, null images (images without annotations) can be included without requiring empty label files, and the suggested ratio is up to 10%. However, in YOLOv8 and later versions, empty label files are recommended for null images to ensure they are properly processed during training. For optimal ratios, 0-10% null images are generally effective for reducing false positives in detection tasks.

2. To verify behavior regarding null images in the code, you can inspect the dataset loading logic in the `load_labels()` function within the Ultralytics source. This typically determines whether images without annotations are skipped or included. For YOLOv8, explicitly adding empty label files ensures they are utilized as intended.

3. Yes, including null images can also help in segmentation tasks by teaching the model to identify when no objects are present. In your bin-picking scenario, providing such null images with empty annotations can help the model learn cases where segmentation is not possible due to overlapping objects.

Let us know if you have further questions!

**Comment by pvtoan** [comment]
Hi @glenn-jocher, 

Thank you for your clear explanations!

Related to empty label files for segmentation task, is an empty file (*.json) good enough?

Or, only "points" is empty while other parts including flags, shapes, imageHeight, imageWidth, ... are still required in that *.json file?

**Comment by glenn-jocher** [comment]
@pvtoan hello,  

For segmentation tasks, an empty annotation file (e.g., a JSON file) should still include the necessary metadata such as `"flags"`, `"shapes"`, `"imageHeight"`, and `"imageWidth"`. Only the `"points"` field can be left empty to indicate the absence of labeled objects. This ensures the file is properly structured and compatible with the dataset loader.  

Let us know if you have further questions!

**Comment by pvtoan** [comment]
Hi @glenn-jocher ,

I got it. 

Thank you!

**Comment by glenn-jocher** [comment]
You're welcome! Let us know if you have any further questions.

**Comment by hengamirah** [comment]
> In summary, it's good practice to label/annotate _every_ instance of objects you aim to use your model to detect with. Failing to do so will result in a poorly trained model as unlabeled data is considered "background" or "null" samples. There are caveats and exceptions to this that I would consider edge-cases, and I would advise that **all objects are labeled. Always.**

![image](https://github.com/user-attachments/assets/579b3677-0dea-41db-bd95-89a757f78fe6)

**Comment by glenn-jocher** [comment]
@hengamirah thank you for your question. To use images with YOLO, you need to label them according to the task you're performing. For object detection, annotate objects with bounding boxes and their respective class labels. If you're transitioning from a model like ResNet50, which typically performs classification, you may need to switch to object detection-specific annotations. Tools like [LabelImg](https://github.com/HumanSignal/labelImg) or [Roboflow Annotate](https://roboflow.com/annotate?ref=ultralytics) can help simplify this process. Make sure your annotations align with YOLO's format for optimal model training. Let me know if you need further clarification!

**Comment by felixkarevo** [comment]
Hi everyone! I am interested to know in which set to place null images. From what I read above I would assume that they should be in both in the training and the validation set? Is that correct? 

My first intuition would have been to place them in the validation set only as there is nothing to "learn" from in null images that have empty labels. However, it seems my intuition is wrong. Apparently the model needs to see null images during training (i.e. in the training set) to learn to reject false positives effectively, right? Which means a model does not only learn from areas/images that are labeled but actually from the entire image not matter if an area is labeled or not.

In short: Null images should be in both the validation and training set, right?

**Comment by kv-22** [comment]
I have null class images under the unannotated section of the dataset. How can I add these images to the dataset as null class?

**Comment by glenn-jocher** [comment]
@kv-22 To add your unannotated images as null class samples to your dataset:

1. Create an empty text file (with .txt extension) for each image with the same filename as the image
2. Place these empty txt files in your labels directory, matching the structure of your annotated data
3. Make sure your dataset configuration (YAML file) points to both the images and the directory containing these empty label files

These empty annotation files signal to the dataset loader that these images should be included during training as valid samples with no objects, rather than being filtered out. This approach helps your model learn to reject false positives in scenarios where no objects are present.

**Comment by kv-22** [comment]
> [@kv-22](https://github.com/kv-22) To add your unannotated images as null class samples to your dataset:
> 
> 1. Create an empty text file (with .txt extension) for each image with the same filename as the image
> 2. Place these empty txt files in your labels directory, matching the structure of your annotated data
> 3. Make sure your dataset configuration (YAML file) points to both the images and the directory containing these empty label files
> 
> These empty annotation files signal to the dataset loader that these images should be included during training as valid samples with no objects, rather than being filtered out. This approach helps your model learn to reject false positives in scenarios where no objects are present.

Thank you for the reply, however, I wasn't able to make it work. When I upload the image and the empty txt file, it automatically gets detected as unannotated. All the images go under the unannotated section. Also, from where can I access the txt or yaml files in the workspace?

**Comment by glenn-jocher** [comment]
@kv-22 It sounds like you might be using Ultralytics HUB rather than a local file system. In HUB, images with empty annotation files are still categorized as "unannotated" by design. 

To properly include null images in your dataset:

1. If using local training: Create empty .txt files in your labels directory and reference them in your dataset.yaml file
2. If using Ultralytics HUB: You'll need to export your dataset, add the empty annotation files locally, then re-import, or use the API to programmatically manage these files

For accessing YAML files in HUB, you can download your dataset using the export option, modify it locally, then re-upload. Alternatively, you can define the dataset structure directly in your training script.

Would you mind clarifying which environment you're working in so I can provide more specific guidance?

**Comment by kv-22** [comment]
> [@kv-22](https://github.com/kv-22) It sounds like you might be using Ultralytics HUB rather than a local file system. In HUB, images with empty annotation files are still categorized as "unannotated" by design.
> 
> To properly include null images in your dataset:
> 
> 1. If using local training: Create empty .txt files in your labels directory and reference them in your dataset.yaml file
> 2. If using Ultralytics HUB: You'll need to export your dataset, add the empty annotation files locally, then re-import, or use the API to programmatically manage these files
> 
> For accessing YAML files in HUB, you can download your dataset using the export option, modify it locally, then re-upload. Alternatively, you can define the dataset structure directly in your training script.
> 
> Would you mind clarifying which environment you're working in so I can provide more specific guidance?

Sorry for not specifying it earlier, I'm actually putting the dataset in Roboflow. I would appreciate it if you have any knowledge of working in this environment, otherwise, I will check the rest of the options you've mentioned, thank you.

**Comment by glenn-jocher** [comment]
@kv-22 Thanks for clarifying that you're using Roboflow. For including null images in Roboflow:

1. In Roboflow, you can add images without annotations by uploading them to your project
2. During the "Generate" step, select the option to "Include images with no annotations" in the preprocessing settings
3. This will ensure these empty images are included in your exported dataset

Alternatively, you can use Roboflow's annotation tools to create empty annotation files by simply not drawing any bounding boxes on those images, and then manually marking them as "Annotated" in the interface.

If you continue to have issues, Roboflow's documentation and support team might provide platform-specific guidance since they handle dataset preparation differently than our direct YOLO implementation.

**Comment by 1ouOne** [comment]
Hi @glenn-jocher ! I have almost same question as @felixkarevo .
I have less than 10% of my whole dataset as images with target objects but without annotations (empty txt files) and I spread them randomly across all splits. Can I include such images in my validation set or I should distribute them across train and test splits only?
What's the standard practice for handling such images in dataset splits?

**Comment by glenn-jocher** [comment]
@1ouOne Yes, you should include null images in your validation set as well as your training set. This allows you to properly evaluate your model's ability to reject false positives during validation, which is crucial for assessing real-world performance. Your approach of randomly distributing them across all splits (train/val/test) is the standard practice and will give you the most representative evaluation of your model's performance.

**Comment by 1ouOne** [comment]
Thank you a lot @glenn-jocher !
You've made me sleep easier this night!

**Comment by glenn-jocher** [comment]
@1ouOne You're very welcome! Glad I could help clarify that for you. Sweet dreams and good luck with your model training! 😊

**Comment by asudeeaydin** [comment]
Hi @glenn-jocher, 

I've read the thread a couple of times now. But I still feel a bit lost, I'm using a Yolov6 network and have built my own dataloader. Could you describe what's going on from a lower level as well? ie.

What are the bbox and class values passed in for the absence of a label (ie. a null image), are they None values?
What is the loss score exactly and how a null image contributes to ie (ie. is it only through the objectiveness score).
I've tried to use None values in my custom dataloader but soon realized that this fails in the collate_fn function since None values are not a thing in torch.

**Comment by glenn-jocher** [comment]
@asudeeaydin For null images in YOLO implementations, you should pass empty tensors rather than None values:

1. **Bbox and class values**: Pass empty tensors with the correct shape - `torch.zeros((0, 4))` for bboxes and `torch.zeros(0)` for classes. This maintains tensor structure while indicating no objects.

2. **Loss contribution**: Null images primarily contribute through the objectness/confidence loss component. The model learns to predict low confidence scores across all anchor points when no objects are present, helping reduce false positives.

3. **Collate_fn solution**: Instead of None values, structure your dataloader to return empty tensors. For example:
```python
if no_objects:
    return {
        'bboxes': torch.zeros((0, 4)), 
        'classes': torch.zeros(0, dtype=torch.long),
        'image': image_tensor
    }
```

This approach ensures the tensors can be properly batched in collate_fn while still representing the absence of objects. The model's objectness predictions will be penalized during training when they incorrectly predict high confidence on these null samples.
